### PR TITLE
[Improvements] A bit of everything

### PR DESCRIPTION
**Pull request by Zoma-Ancestral**
PR done to regroup multiple small changes and improvements mainly caused by W6 and subsequent patches until now

For now in this PR :

Farming :
- Plots :
    - Now won't display OG if OG isn't unlocked yet
    - Now display the star icon for undiscovered crops at all times (and not only while growing)
    - Now actualize correctly in the UI (if you had a grown plot, collect it then trigger a cloud save without leaving the plot page, the timer would get to the right time but all would be displayed as if it's still an OG cycle)
- Market upgrades now display cost to the max level or a certain threshold if the max level is too far
- Market upgrades now try to check if you can buy next level with crops to collect or beans to exchange, if yes display the cost in orange

Summoning :
- New Sailing artifact The Winz Lantern is now included in the Winner bonuses calculation

Deathnote :
-  W6 monsters have been added to Deathnote
- Minibosses from Jade Upgrade have been added

Equinox :
- New alert if the bar is full

Star Signs :
- Star Signs bonus gotten from Infinity Rift bonus now won't be obtained if Star Sign isn't unlocked first
- Star Signs bonus now takes into account the Seraph Cosmos bonus if unlocked

Breeding :
- Small fix, all pets in the fence yard are now correctly retrieved to highlight shinies currently in it (was missing the last one)

Cooking :
- Cooking speed now takes into account all relevant bonuses and is now accurate
- Kitchen upgrade cost now takes into account all relevant bonuses and is now accurate
- Maximum meal level now takes into account new bonuses and can correctly reach 90
- In the UI, Time to 60 has been replaced by Time to Max, which is calculated depending on the max level for the account
- Now display which meal will be upgraded by the No Meal Left Behind Emporium bonus

Printer :
- Lolly Flower pristine charm is now included in printer output calculation

Stamps :
- Liqorice Rolle pristine charm is now included in stamps (except misc) bonuses calculation

Players :
- Gumball Necklace pristine charm is now included in cash from monsters calculation
- Bean stalking bonus is now included in the cash from monster calculation
- Updated bonus to Golden Food effect for the cash from monster calculation
- New Sailing artifact The Onyx Lantern is now included in the Onyx statues bonus calculation
- UI page for talents:
    - Now display correctly all sources of increased talent level
    - "Show book level" is now accurate again
    - Talents that have been maxed (level = max level) are now highlighted

Anvil :
- Bean stalking bonus is now included in All Skill Exp
- Updated bonus to Golden Food effect for All Skill Exp calculation

Sneaking : 
- Added a basic page for Beanstalk once unlocked to keep track of unlocked bonuses (only non-boosted values displayed as Gold Food boost effect is very character dependant)

Gaming :
- Nugget size range now takes into account the Metal Detector Equinox bonus
- Shovel dig number available calculation has been updated (fractal island bonus is still missing as it is not supported in IE yet)
- Shovel time until the next dig has been updated to be in adequation with the Shovel dig available calculation

Slab :
- Now display Slab bonuses on the Slab page
- New Lab node Slab Sovereignty is now applied to slab bonuses

Alchemy :
- Now display the total bonus for bubbles that benefit from Slab
- Sigils :
    - Now support Ionized versions
    - Updated UI to display charge speed per player

Lab :
- Now support new Lab nodes, they'll be greyed out if not purchased, and can't be activated until they're bought at the Emporium
- Now include the Summoning Winner bonus into the bonus connection range calculation
- Now support new Lab Jewels

Sailing :
- Artifacts :
    - Now support Sovereign rank
    - The 3 new artifacts are now supported
- Sailing speed can now take the C. Shanti Minor star sign into the calculation
- 2hrs dist is now Ideal dist and displays the distance covered by the boat within the Minimal Travel Time

Gods :
- Now take into account the Emporium bonus to blessings

Worship :
- Worship charge speed and max charge are accurate again
- Totalizer bonuses display has been moved to the Totem tab

Stamps :
- Unlocked stamps that are still level 0 aren't greyed out anymore
- Fix a bug where refreshing the site while being on the Stamp page set the displayed total stamp levels to 0

Silkrode checkbox :
Some global pages (so not linked to a specific player) are now in need of using a star sign to be accurate.
To avoid having some weird player tracking and for consistency's sake, the concerned pages now have multiple checkboxes.
For each star sign in the calculations, there'll be a checkbox to consider whether the sign is equipped or not (force-checked if the sign is affected by Infinite Stars Rift bonus), and a last checkbox to apply Silkrode Chip (double the equipped star signs effects).
For now, there are only 3 pages with this system: Cooking, Sailing, and Farming

TODO in a future PR :
- [ ] Add support for W2 Killroy battles
- [ ] Add support for W2 Expeditions Islands
- [ ] Add support for The Tome

### CODE REVIEW

**Comment by Zoma-Ancestral** [comment]
As previous times, PR is still being worked on so shouldn't be merged for now

**Comment by Zoma-Ancestral** [comment]
I'm done with the PR, waiting for the review and the possible adjustments to do

File: data/domain/player.tsx [hunk 1]
```diff
@@ -708,11 +738,86 @@ export const updatePlayerDeathnote = (data: Map<string, any>) => {
...
+}
+
+export const updatePlayerTalentLevelExceptESBonus = (data: Map<string, any>) => {
+    const players = data.get("players") as Player[];
+    const divinity = data.get("divinity") as Divinity;
+    const equinox = data.get("equinox") as Equinox;
+    const achievementsInfo = data.get("achievements") as Achievement[];
+
+    // Update max talent level due to bear.
+    // Update players talents levels due to elite class level increase talents.
```
Comment (Sludging) [inline data/domain/player.tsx:769]: You copied my comment maybe but if I'm not wrong it's only bear here? I don't remember where the symbols of beyond stuff comes from 😆

**Review by Sludging — APPROVED** [review]
I'll be honest, I gave up half way through.

Please don't make PRs of this kind again if you can help it, it's impossible to review and it's very difficult to give you suggestions / recommendations / feedback like this.

As I'm away, I can't verify any of the math here / fixes are correct. So I'm going to put this PR on the https://preview.idleonefficiency.com/ domain and ask for people in the discord to test it out and verify numbers are correct.

Regardless of the above statements, really amazing work on getting so much done! I can tell a lot of effort has gone into this and I really really appreciate it.

I'm going to approve it but please wait with merging until we get 1-2 people commenting about the accuracy of the numbers.

File: components/world-6/farming/marketUpgrades.tsx
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply components/world-6/farming/marketUpgrades.tsx:1]: Ok I see, next time as PRs will have fewer changes it'll be easier to ask for your opinion on those kinds of things, I'll keep that in mind

File: components/world-6/sneaking/beanstacking.tsx
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply components/world-6/sneaking/beanstacking.tsx:1]: Nice idea for improvement indeed

File: data/domain/alchemy.tsx [hunk 2]
```diff
@@ -168,6 +169,28 @@ export class Bubble {
...
+    // Math.floor(this.lootyCount / 100)
+    override getBonus = (roundResult: boolean = false): number => {
+        const bonus = lavaFunc(this.func, this.level, this.x1, this.x2, false);
+        return roundResult ? round(Math.floor(this.lootyCount / 100) * bonus) : Math.floor(this.lootyCount / 100) * bonus;
+    }
+
+    override getBonusText = (bonus: number = this.getBonus(true)): string => {
+        let titleText = this.description.replace(/{/g, lavaFunc(this.func, this.level, this.x1, this.x2, true).toString());
+        titleText += ` (${this.lootyCount} items found = ${bonus.toString()} bonus stats)`;
+        return handleToolBubbles(titleText, this.name);
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/alchemy.tsx:190]: Yeah, I copy-pasted the override getBonusText from the Diamond bubble I think, or another special bubble to make it, so I kept it as is, but will change that to just return basic string

File: data/domain/alchemy.tsx [hunk 3]
```diff
@@ -329,7 +355,7 @@ export class Vial {
     }
 
     getNumberToRoll = () => {
-        return 100 - vialPercentages[this.vialIndex];
+        return 100 - vialPercentages[this.vialIndex < vialPercentages.length ? this.vialIndex : vialPercentages.length - 1];
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/alchemy.tsx:358]: Gonna update the array

File: data/domain/alerts.tsx [hunk 4]
```diff
@@ -325,6 +335,12 @@ const getGlobalAlerts = (worship: Worship, refinery: Refinery, traps: Trap[][],
         globalAlerts.push(new FoodLustAlert());
     }
 
+    // Equinox Bar Full
+    const equinoxBarFull = (equinox.bar.current == equinox.bar.max);
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/alerts.tsx:339]: Done

File: data/domain/alerts.tsx [hunk 5]
```diff
@@ -194,6 +195,15 @@ export class CropReadyAlert extends GlobalAlert {
     }
 }
 
+export class EquinoxBarFull extends GlobalAlert {
+    constructor() {
+        super(`Equinox bar is full, go upgrade a cloud`, AlertType.EquinoxFull, Equinox.cloudImageData());
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/alerts.tsx:200]: Indeed you're right x)

File: data/domain/cooking.tsx [hunk 6]
```diff
@@ -277,17 +324,30 @@ export class Cooking extends Domain {
...
     foodLustDiscountCapped: boolean = false;
 
+    starSignUnlocked: boolean = false;
+    starSignInfinity: boolean = false;
+
     getMaxMeals = () => {
         return this.meals.length;
     }
 
+    getTotalCookingSpeed = (starSignEquipped: boolean, silkrodeBonus: boolean) => {
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/cooking.tsx:343]: Yeah, one is inside the meal to use for calculations, and the other one is inside the Cooking domain to display it

First planned to use the one in the Cooking Domain inside Meals but it added another parameter to each function so it wasn't really clean into the page code, so I preferred to duplicate it inside the Domain.

Would you prefer me to change that?

File: data/domain/cooking.tsx [hunk 7]
```diff
@@ -531,74 +617,63 @@ export const updateCooking = (data: Map<string, any>) => {
...
+        kitchen.mealSpeedWithoutStarSign = kitchen.getMealSpeed(0, mealCookVialBonus, fireflyVialBonus, turtleVialBonus, winnerBonus, stampBonus, meal63Bonus, mealSpeedBonus, jewel0Bonus, trollCardBonus, ceramicCardBonus, kitchenEfficientBonus, jewel14Bonus, diamonChef, achievements[225].completed, achievements[224].completed, atomBonus, artifactBonus, worshipBonus, bloodMarrowBonus, superChowBonus, cropScientistBonus, farming.farmingLevel, arcadeBonus);
+        kitchen.mealSpeedWithSilkrode = kitchen.getMealSpeed(starsign58*2, mealCookVialBonus, fireflyVialBonus, turtleVialBonus, winnerBonus, stampBonus, meal63Bonus, mealSpeedBonus, jewel0Bonus, trollCardBonus, ceramicCardBonus, kitchenEfficientBonus, jewel14Bonus, diamonChef, achievements[225].completed, achievements[224].completed, atomBonus, artifactBonus, worshipBonus, bloodMarrowBonus, superChowBonus, cropScientistBonus, farming.farmingLevel, arcadeBonus);
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/cooking.tsx:628]: Yeah I know, tried to find another way to do it more cleanly but didn't though about anything either '-'
But I think it's still cleaner than having 3 whole blocks of recalculated values for each type

File: data/domain/gaming.tsx [hunk 8]
```diff
@@ -64,6 +68,12 @@ export class Gaming extends Domain {
     rawGamingData: any[] = [];
     rawSproutData: number[][] = [];
 
+    equinoxBonustoNuggets: number = 1;
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/gaming.tsx]: Updated this, small typo indeed x)

File: data/domain/lab.tsx [hunk 9]
```diff
@@ -40,6 +43,9 @@ export class MainframeBonus {
         this.bonusOff = data.bonusOn;
         this.name = data.name;
         this.description = data.description;
+        if (index < 14) {
+            this.unlocked = true;
+        }
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/lab.tsx:49]: Added a comment there

File: data/domain/lab.tsx [hunk 10]
```diff
@@ -543,14 +604,22 @@ export const updateLab = (data: Map<string, any>) => {
...
+        (lab.bonuses[9] as FungiFingerBonus).jewelBoost = lab.jewels[13].getBonus();
     }
     (lab.bonuses[11] as UnadulteratedBankingBonus).greenStacks = [...new Set(storage.chest.filter(item => item.count >= 1e7).map(item => item.internalName))].length;
 
+    // Emerald Rhombol
+    if (lab.jewels[18].active) {
+        (lab.bonuses[15] as SlabSovereigntyBonus).jewelBoost = lab.jewels[18].getBonus();
+    }
+
+    // Emerald Rhombol
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/lab.tsx]: Yeah copy-pasted code and forgot to adjust the names, fixed

File: data/domain/lab.tsx [hunk 11]
```diff
@@ -543,14 +604,22 @@ export const updateLab = (data: Map<string, any>) => {
     // Special Bonus handling
     (lab.bonuses[0] as AnimalFarmBonus).totalSpecies = breeding.speciesUnlocks.reduce((sum, world) => sum += world, 0);
 
-    (lab.bonuses[9] as FungiFingerBonus).greenMushroomKilled = deathnote.mobKillCount.get("mushG")?.reduce((sum, killCount) => sum += Math.round(killCount), 0) ?? 0;
+    (lab.bonuses.find(bonus => bonus.index == 9) as FungiFingerBonus).greenMushroomKilled = deathnote.mobKillCount.get("mushG")?.reduce((sum, killCount) => sum += Math.round(killCount), 0) ?? 0;
```
[the remark this answers is not in the document]
Comment (Zoma-Ancestral) [reply data/domain/lab.tsx]: I forgot to revert this now that bonuses are sorted, fixed

File: data/domain/player.tsx (hunk 1 above)
Comment (Zoma-Ancestral) [reply data/domain/player.tsx:769]: Yeah indeed x)
Symbols of beyond are the talents, it's done in the parseTalents function

Removed the unnecessary comment line

**Comment by Zoma-Ancestral** [comment]
Addressed all change requests and fixed all problems reported in Discord until now

File: data/domain/cooking.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Sludging) [reply data/domain/cooking.tsx:343]: So are they both still used? I didn't follow all the code.

If only one is used, I'd rather keep only one. If both are used then it's fine, I missed that it's two different classes

**Review by Sludging — APPROVED** [review]

**Pull request merged**